### PR TITLE
feat(api): comment tagging & edit + create comment from editor

### PR DESCRIPTION
**Pull request by Itrulia**
```graphql
mutation createTag($tag:String,$type:TagType!) {
	createTag(tag:$tag,type:$type){
		id
		tag
        type
	}
}

// vars
{
  "tag":"Super Like2",
  "type":"Comment"
}
```

```graphql
mutation deleteTag($id:ID!) {
	deleteTag(id:$id){
		id
		tag
        type
	}
}

// vars
{
  "id":"abcd"
}
```

```graphql
mutation updateTag($id:ID!,$tag:String) {
	updateTag(id:$id,tag:$tag){
		id
		tag
        type
	}
}

// vars
{
  "id":"abcd",
  "tag":"Super Like",
}
```

### CODE REVIEW

File: packages/api/src/graphql/permissions.ts [hunk 1]
```diff
@@ -483,6 +483,30 @@ export const CanUpdateSettings: Permission = {
...
+}
+
+export const CanUpdateTag: Permission = {
+  id: 'CAN_UPDATE_TAG',
+  description: 'Allows to update a tag',
+  deprecated: false
+}
+
+export const CanGetTag: Permission = {
+  id: 'CAN_GET_TAG',
```
Comment (michael-scheurer) [inline packages/api/src/graphql/permissions.ts:499]: Should probably be "CAN_GET_TAGS", so plural. In terms of consistency with other permissions.

File: packages/api/src/graphql/query.private.ts [hunk 2]
```diff
@@ -638,6 +640,23 @@ export const GraphQLQuery = new GraphQLObjectType<undefined, Context>({
...
+      args: {
+        cursor: {type: GraphQLID},
+        take: {type: GraphQLInt, defaultValue: 10},
+        skip: {type: GraphQLInt, defaultValue: 0},
+        filter: {type: GraphQLTagFilter},
+        sort: {type: GraphQLTagSort, defaultValue: TagSort.CreatedAt},
+        order: {type: GraphQLSortOrder, defaultValue: SortOrder.Descending}
+      },
+      resolve: (root, {filter, sort, order, cursor, take, skip}, {authenticate, prisma}) =>
+        getTags(filter, sort, order, cursor, skip, take, authenticate, prisma.tag)
```
Comment (michael-scheurer) [inline packages/api/src/graphql/query.private.ts:659]: I would use object as parameter ;)

File: packages/api/src/graphql/tag/tag.private-query.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    tag.findMany({
+      where: where,
+      skip: skip,
+      take: Math.min(take, MaxResultsPerPage) + 1,
+      orderBy: orderBy,
+      cursor: cursorId ? {id: cursorId} : undefined
+    })
+  ])
+
+  const nodes = tags.slice(0, take)
```
Comment (michael-scheurer) [inline packages/api/src/graphql/tag/tag.private-query.ts:78]: Why is this necessary when you have the take property passed in the findMany() function?

File: packages/api/src/graphql/tag/tag.private-query.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  const where = createTagFilter(filter)
+
+  const [totalCount, tags] = await Promise.all([
+    tag.count({
+      where: where,
+      orderBy: orderBy
+    }),
+    tag.findMany({
+      where: where,
+      skip: skip,
```
Comment (michael-scheurer) [inline packages/api/src/graphql/tag/tag.private-query.ts:71]: just skip

File: packages/api/src/graphql/tag/tag.private-query.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  const orderBy = createTagOrder(sortedField, getSortOrder(order))
+  const where = createTagFilter(filter)
+
+  const [totalCount, tags] = await Promise.all([
+    tag.count({
+      where: where,
+      orderBy: orderBy
+    }),
+    tag.findMany({
+      where: where,
```
Comment (michael-scheurer) [inline packages/api/src/graphql/tag/tag.private-query.ts:70]: just where

File: packages/api/src/graphql/tag/tag.private-query.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      skip: skip,
+      take: Math.min(take, MaxResultsPerPage) + 1,
+      orderBy: orderBy,
```
Comment (michael-scheurer) [inline packages/api/src/graphql/tag/tag.private-query.ts:73]: simplify

**Review by michael-scheurer — DISMISSED** [review]
Only minimal remarks =)

File: packages/api/src/graphql/tag/tag.private-query.ts (hunk 3 above)
Comment (Itrulia) [reply packages/api/src/graphql/tag/tag.private-query.ts:78]: Because with cursor based navigation we do not know if there's another page unfortunately

File: packages/api/src/graphql/query.private.ts (hunk 2 above)
Comment (Itrulia) [reply packages/api/src/graphql/query.private.ts:659]: We currently have this everywhere from the prisma refactor, we should change that all at once for consistency sake in a separate PR

File: package.json [hunk 7]
```diff
@@ -117,6 +117,7 @@
       "prettier/standard"
     ],
     "rules": {
+      "object-shorthand": "warn",
```
Comment (Itrulia) [inline package.json:120]: Automatically turns `{error: error}` into `{error}` with --fix

File: packages/api/src/graphql/tag/tag.private-query.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    case TagSort.CreatedAt:
+    default:
+      return {
+        createdAt: sortOrder
+      }
+  }
+}
+
+const createTypeFilter = (filter?: Partial<TagFilter>): Prisma.TagWhereInput => {
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api/src/graphql/tag/tag.private-query.ts:41]: It's optional on graphql side (like the other filters) so that you don't always have to specify an empty filter on the client.
